pulsar-broker-auth-oidc/src/main/java/org/apache/pulsar/broker/authentication/oidc/OpenIDProviderMetadataCache.java

```diff
@@ -0,0 +1,124 @@
+import static org.apache.pulsar.broker.authentication.oidc.ConfigUtils.getConfigValueAsInt;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.ObjectReader;
+import com.github.benmanes.caffeine.cache.AsyncCacheLoader;
+import com.github.benmanes.caffeine.cache.AsyncLoadingCache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nonnull;
+import javax.naming.AuthenticationException;
+import org.apache.pulsar.broker.ServiceConfiguration;
+import org.asynchttpclient.AsyncHttpClient;
+
+/**
+ * Class used to cache metadata responses from OpenID Providers.
+ */
+class OpenIDProviderMetadataCache {
+
+    private final ObjectReader reader = new ObjectMapper().readerFor(OpenIDProviderMetadata.class);
+
+    /**
+     * A loader for the cache that retrieves the metadata from the issuer's /.well-known/openid-configuration endpoint.
+     * @return a connection to the issuer's /.well-known/openid-configuration endpoint
+     * @throws AuthenticationException if the URL is malformed or there is an exception while opening the connection
+     */
+    private AsyncCacheLoader<String, OpenIDProviderMetadata> getLoader(AsyncHttpClient client) {
+        return (issuer, executor) ->
+                // TODO OIDC spec https://openid.net/specs/openid-connect-discovery-1_0.html#NormalizationSteps
+                // calls for normalization according to RFC3986. Is that important to verify here?
+                client
+                    .prepareGet(issuer + "/.well-known/openid-configuration")
+                    .execute()
+                    .toCompletableFuture()
+                    .thenCompose(result -> {
+                        CompletableFuture<OpenIDProviderMetadata> future = new CompletableFuture<>();
+                        try {
+                            OpenIDProviderMetadata openIDProviderMetadata =
+                                    reader.readValue(result.getResponseBodyAsBytes());
+                            verifyIssuer(issuer, openIDProviderMetadata);
+                            future.complete(openIDProviderMetadata);
+                        } catch (AuthenticationException e) {
+                            incrementFailureMetric(AuthenticationExceptionCode.ERROR_RETRIEVING_PROVIDER_METADATA);
+                            future.completeExceptionally(e);
+                        } catch (Exception e) {
+                            incrementFailureMetric(AuthenticationExceptionCode.ERROR_RETRIEVING_PROVIDER_METADATA);
+                            future.completeExceptionally(new AuthenticationException(
+                                    "Error retrieving OpenID Provider Metadata at " + issuer + ": " + e.getMessage()));
+                        }
+                        return future;
```

Comment:
By whatever means, please use the actual value of `OpenIDProviderMetadata::getIssuer()` as the key into the cache, as opposed to the configured issuer value.

Reply:
I modified the cache a bit to suit our needs more. It probably still needs tweaking though.